### ISSUE TITLE
Non-Termination for Witnesses of Recursive Types 

### ISSUE THREAD

**Issue by Qbxs** [Qbxs/qualified-subtyping#3]
Even though subconstraints of recursive types eventually lead to a `cacheHit`, the algorithm won't terminate in such cases...

**Comment by Qbxs** [comment]
The (commented out) example in `main` eventually leads to a `cacheHit` with the following subconstraint and  `SolverState`:

```Haskell
-- State on cacheHit in example
-- This should lead to no subconstraints being added, i.e. termination

-- constraint that is generated again:
subconstr = Subtype (FuncTy Nat (RecTy "a" (FuncTy Nat (RecVar "a")))) (FuncTy Nat (RecTy "a" (FuncTy Nat (RecVar "a"))))

-- state:
state = SolverState
           { ss_cache = fromList [ (Subtype Bot Top,'a')
                                 , (Subtype (FuncTy Nat (RecTy "a" (FuncTy Nat (RecVar "a")))) (FuncTy Nat (RecTy "a" (FuncTy Nat (RecVar "a")))),'c')
                                 , (Subtype (FuncTy Nat (RecTy "a" (FuncTy Nat (RecVar "a")))) (RecTy "a" (FuncTy Nat (RecVar "a"))),'f')
                                 , (Subtype Nat Nat,'d')
                                 , (Subtype (RecTy "a" (FuncTy Nat (RecVar "a"))) (FuncTy Nat (RecTy "a" (FuncTy Nat (RecVar "a")))),'b')
                                 , (Subtype (RecTy "a" (FuncTy Nat (RecVar "a"))) (RecTy "a" (FuncTy Nat (RecVar "a"))),'e')]
            , ss_known = fromList [ ('a',FromTop Bot)
                                  , ('b',Rec (SubVar 'c'))
                                  , ('c',Func (SubVar 'd') (SubVar 'e'))
                                  , ('d',Refl Nat)
                                  , ('e',Rec (SubVar 'f'))]
            , ss_fresh = ['a'..]
            }
```

**Comment by Qbxs** [comment]
The issue was a mutual reference of witness variables in `substitute` which has been fixed in 971c8b4c2e043141437bc5c4202a1ce185d2d3f2
